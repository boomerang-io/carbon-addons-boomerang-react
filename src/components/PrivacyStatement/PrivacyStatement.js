import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import dompurify from 'dompurify';
import { Accordion, AccordionItem, Button } from 'carbon-components-react';
import {
  ModalHeader,
  ModalBody,
  ModalFooter,
} from 'carbon-components-react/lib/components/ComposedModal';
import { settings } from 'carbon-components';
import window from 'window-or-global';

import HeaderMenuModalWrapper from '../../internal/HeaderMenuModalWrapper';
import HeaderMenuItem from '../HeaderMenuItem';
import notify from '../Notifications/notify';
import ToastNotification from '../Notifications/ToastNotification';

import CastleImg from './assets/x_castle_drawbridge.js';

const { prefix } = settings;

/**
 * Helper to format date timestamp
 * @param {string} timestamp
 * @returns formatted date string
 */
function formatDateTimestamp(timestamp) {
  return new Date(timestamp).toLocaleDateString('en-us', {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  });
}

PrivacyStatement.propTypes = {
  baseServiceUrl: PropTypes.string.isRequired,
};

function PrivacyStatement({ baseServiceUrl }) {
  const [statement, setStatement] = useState();
  const [error, setError] = useState();
  const [alertError, setAlertError] = useState();

  useEffect(() => {
    axios(`${baseServiceUrl}/users/consents`)
      .then((response) => setStatement(response.data))
      .catch((err) => setError(err));
  }, [baseServiceUrl]);

  function handleSubmit({ closeAlertModal, closeModal }) {
    axios
      .put(`${baseServiceUrl}/users/consent`, {
        version: statement.version,
        hasConsented: false,
      })
      .then(() => {
        notify(
          <ToastNotification
            subtitle="Successfully requested account deletion"
            title="Delete Account"
            kind="success"
          />,
          { containerId: `${prefix}--bmrg-header-notifications` }
        );
        // want to clear out any error state
        setError(false);
        setAlertError(false);
        closeAlertModal();
        closeModal();
        if (window.location) {
          window.location.reload(true);
        }
      })
      .catch((err) => {
        setAlertError(err);
        closeAlertModal();
      });
  }

  // TOOD: decide to do something if there is an error or not
  if (error) {
    // noop
  }

  return (
    <HeaderMenuItem
      className={`${prefix}--bmrg-privacy-statement-container`}
      iconName="locked"
      text="Privacy Statement"
    >
      {({ closeModal }) => {
        return (
          <>
            <ModalHeader
              closeModal={() => {
                closeModal();
                setError('');
                setAlertError('');
              }}
              label={`Effective as of ${
                statement ? formatDateTimestamp(statement.effectiveDate) : ''
              }`}
              title="Privacy Statement"
            />

            <ModalBody>
              <div className={`${prefix}--bmrg-privacy-statement`}>
                {statement && statement.formContent && statement.formContent.sections.length > 0 && (
                  <Accordion>
                    {statement.formContent.sections.map((section) => {
                      return (
                        <AccordionItem title={section.title} key={section.title}>
                          <p
                            className={`${prefix}--bmrg-privacy-statement__content`}
                            dangerouslySetInnerHTML={{
                              __html: dompurify.sanitize(section.content),
                            }}
                          />
                        </AccordionItem>
                      );
                    })}
                  </Accordion>
                )}
                <p className={`${prefix}--bmrg-privacy-statement__message`}>
                  For any questions or concerns about business and personal information captured on
                  IBM Boomerang, please contact{' '}
                  <a href="mailto:boomrng@us.ibm.com?subject=Boomerang Privacy Statement">
                    boomrng@us.ibm.com
                  </a>
                  .
                </p>
                {alertError && (
                  <p className={`${prefix}--bmrg-privacy-statement__error`}>
                    Failed to recieve deletion request. Please try again.
                  </p>
                )}
              </div>
            </ModalBody>
            <ModalFooter>
              <div className={`${prefix}--bmrg-privacy-statement-delete`}>
                <HeaderMenuModalWrapper
                  buttonTriggerText="Request account deletion"
                  triggerButtonKind="danger"
                >
                  {({ closeModal: closeAlertModal }) => {
                    return (
                      <>
                        <ModalHeader
                          closeModal={closeAlertModal}
                          // label="Delete Account"
                          title="Request account deletion"
                        />
                        <ModalBody>
                          <>
                            <CastleImg />
                            <p className={`${prefix}--bmrg-privacy-statement-delete__desc`}>
                              We will happily delete your account and all corresponding data from
                              our systems. Your account will cease to exist, and we will notify your
                              teams that you are no longer a member of Boomerang. This process can
                              take up to 1 month, which is in accordance with GDPR.
                            </p>
                          </>
                        </ModalBody>
                        <ModalFooter
                          style={{
                            marginTop: '1.125rem',
                            // display: 'flex',
                            // justifyContent: 'space-evenly',
                          }}
                        >
                          <Button kind="secondary" onClick={closeAlertModal}>
                            Cancel
                          </Button>
                          <Button
                            kind="primary"
                            type="submit"
                            onClick={() => {
                              handleSubmit({ closeAlertModal, closeModal });
                            }}
                          >
                            Delete my account
                          </Button>
                        </ModalFooter>
                      </>
                    );
                  }}
                </HeaderMenuModalWrapper>
              </div>
            </ModalFooter>
          </>
        );
      }}
    </HeaderMenuItem>
  );
}

export default PrivacyStatement;
