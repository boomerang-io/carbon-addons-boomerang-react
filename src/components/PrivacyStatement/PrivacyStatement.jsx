import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import dompurify from "dompurify";
import { Accordion, AccordionItem, Button } from "@carbon/react";
import { ModalHeader, ModalBody, ModalFooter } from "@carbon/react";
import { prefix } from "../../internal/settings";
import HeaderMenuModalWrapper from "../../internal/HeaderMenuModalWrapper";
import HeaderMenuItem from "../HeaderMenuItem";
import notify from "../Notifications/notify";
import ToastNotification from "../Notifications/ToastNotification";

import CastleImg from "./assets/x_castle_drawbridge";



/**
 * Helper to format date timestamp
 * @param {string} timestamp
 * @returns formatted date string
 */
function formatDateTimestamp(timestamp) {
  return new Date(timestamp).toLocaleDateString("en-us", {
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}

PrivacyStatement.propTypes = {
  baseServiceUrl: PropTypes.string.isRequired,
  organization: PropTypes.string,
  platformEmail: PropTypes.string,
};

PrivacyStatement.defaultProps = {
  organization: "IBM Services Essentials",
  platformEmail: "isesupp@us.ibm.com",
};

function PrivacyStatement({ baseServiceUrl, organization, platformEmail }) {
  const [statement, setStatement] = useState();
  const [error, setError] = useState();
  const [alertError, setAlertError] = useState();

  useEffect(() => {
    axios(`${baseServiceUrl}/users/consents`)
      .then((response) => setStatement(response.data))
      .catch((err) => setError(err));
  }, [baseServiceUrl]);

  function handleSubmit({ closeAlertModal, closeModal }) {
    axios
      .put(`${baseServiceUrl}/users/consent`, {
        version: statement.version,
        hasConsented: false,
      })
      .then(() => {
        notify(
          <ToastNotification
            subtitle="Successfully requested account deletion"
            title="Delete Account"
            kind="success"
          />,
          { containerId: `${prefix}--bmrg-header-notifications` }
        );
        // want to clear out any error state
        setError(false);
        setAlertError(false);
        closeAlertModal();
        closeModal();
        if (window.location) {
          window.location.reload(true);
        }
      })
      .catch((err) => {
        setAlertError(err);
        closeAlertModal();
      });
  }

  // TOOD: decide to do something if there is an error or not
  if (error) {
    // noop
  }

  return (
    <HeaderMenuItem
      className={`${prefix}--bmrg-privacy-statement-container`}
      iconName="locked"
      text="Privacy Statement"
    >
      {({ closeModal }) => {
        return (
          <>
            <ModalHeader
              closeModal={() => {
                closeModal();
                setError("");
                setAlertError("");
              }}
              label={`Effective as of ${statement ? formatDateTimestamp(statement.effectiveDate) : ""}`}
              title="Privacy Statement"
            />

            <ModalBody>
              <div className={`${prefix}--bmrg-privacy-statement`}>
                {statement && statement.formContent && statement.formContent.sections.length > 0 && (
                  <Accordion>
                    {statement.formContent.sections.map((section) => {
                      return (
                        <AccordionItem title={section.title} key={section.title}>
                          <p
                            className={`${prefix}--bmrg-privacy-statement__content`}
                            dangerouslySetInnerHTML={{
                              __html: dompurify.sanitize(section.content),
                            }}
                          />
                        </AccordionItem>
                      );
                    })}
                  </Accordion>
                )}
                <p className={`${prefix}--bmrg-privacy-statement__message`}>
                  {`For any questions or concerns about business and personal information captured on
                  ${organization}, please contact${" "}`}
                  <a href={`mailto:${platformEmail}?subject=${organization} Privacy Statement`}>{platformEmail}</a>.
                </p>
                {alertError && (
                  <p className={`${prefix}--bmrg-privacy-statement__error`}>
                    Failed to recieve deletion request. Please try again.
                  </p>
                )}
              </div>
            </ModalBody>
            <ModalFooter>
              <div className={`${prefix}--bmrg-privacy-statement-delete`}>
                <HeaderMenuModalWrapper buttonTriggerText="Request account deletion" triggerButtonKind="danger">
                  {({ closeModal: closeAlertModal }) => {
                    return (
                      <>
                        <ModalHeader
                          closeModal={closeAlertModal}
                          // label="Delete Account"
                          title="Request account deletion"
                        />
                        <ModalBody>
                          <>
                            <div className={`${prefix}--bmrg-privacy-statement-delete__castle`}>
                              <CastleImg style={{ width: "100%" }} />
                            </div>
                            <p className={`${prefix}--bmrg-privacy-statement-delete__desc`}>
                              {`We will happily delete your account and all corresponding data from
                              our systems. Your account will cease to exist, and we will notify your
                              teams that you are no longer a member of ${organization}. This process can
                              take up to 1 month, which is in accordance with GDPR.`}
                            </p>
                          </>
                        </ModalBody>
                        <ModalFooter
                          style={{
                            marginTop: "1.125rem",
                            // display: 'flex',
                            // justifyContent: 'space-evenly',
                          }}
                        >
                          <Button kind="secondary" onClick={closeAlertModal}>
                            Cancel
                          </Button>
                          <Button
                            kind="primary"
                            type="submit"
                            onClick={() => {
                              handleSubmit({ closeAlertModal, closeModal });
                            }}
                          >
                            Delete my account
                          </Button>
                        </ModalFooter>
                      </>
                    );
                  }}
                </HeaderMenuModalWrapper>
              </div>
            </ModalFooter>
          </>
        );
      }}
    </HeaderMenuItem>
  );
}

export default PrivacyStatement;
