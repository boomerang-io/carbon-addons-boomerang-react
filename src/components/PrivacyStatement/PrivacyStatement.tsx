import React from "react";
import { Accordion, AccordionItem, Button, ComposedModal, ModalHeader, ModalBody, ModalFooter } from "@carbon/react";
import { Locked } from "@carbon/react/icons";
import dompurify from "dompurify";
import ErrorMessage from "../ErrorMessage";
import HeaderMenuItem from "../Header/HeaderMenuItem";
import ToastNotification from "../Notifications/ToastNotification";
import Loading from "../Loading";
import notify from "../Notifications/notify";
import { useQuery, useMutation } from "react-query";
import { serviceUrl, resolver } from "../../config/servicesConfig";
import { prefix } from "../../internal/settings";

function formatDateTimestamp(timestamp: string) {
  return new Date(timestamp).toLocaleDateString("en-us", {
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}

type Props = {
  baseServicesUrl: string;
  closeModal: () => void;
  isOpen: boolean;
  organization?: string;
  platformEmail?: string;
};

function PrivacyStatement({
  baseServicesUrl,
  closeModal,
  isOpen,
  organization = "the platform",
  platformEmail = "isesupp@us.ibm.com",
}: Props) {
  const [isConfirmModalOpen, setIsConfirmModalOpen] = React.useState(false);
  const statementUrl = serviceUrl.getStatement({ baseServicesUrl });
  const statementQuery = useQuery({
    queryKey: statementUrl,
    queryFn: resolver.query(statementUrl),
  });

  const { mutateAsync, error: mutateUserConsentError } = useMutation(resolver.putUserConsent);

  async function handleSubmit({
    closeAlertModal,
    closeModal,
  }: {
    closeAlertModal: () => void;
    closeModal: () => void;
  }) {
    const body = {
      hasConsented: false,
      version: statementQuery.data.version,
    };

    try {
      await mutateAsync({ baseServicesUrl, body });
      notify(
        <ToastNotification subtitle="Successfully requested account deletion" title="Delete Account" kind="success" />,
        { containerId: `${prefix}--bmrg-header-notifications` }
      );
      closeAlertModal();
      closeModal();
      if (window.location) {
        window.location.reload();
      }
    } catch (e) {
      closeAlertModal();
    }
  }

  return (
      <ComposedModal aria-label="Privacy Statement" open={isOpen} className={`${prefix}--bmrg-privacy-statement-container ${prefix}--bmrg-header-modal`} onClose={closeModal}>
        <ModalHeader
          closeModal={closeModal}
          label={`Effective as of ${statementQuery.data ? formatDateTimestamp(statementQuery.data.effectiveDate) : ""}`}
          title="Privacy Statement"
        />
        <ModalBody>
          <div className={`${prefix}--bmrg-privacy-statement`}>
            {statementQuery.isLoading ? (
              <Loading />
            ) : statementQuery.error ? (
              <ErrorMessage style={{ color: "#F2F4F8" }} />
            ) : (
              statementQuery.data?.formContent?.sections?.length > 0 && (
                <>
                  <Accordion>
                    {statementQuery.data.formContent.sections.map((section: any) => {
                      return (
                        <AccordionItem title={section.title} key={section.title}>
                          <p
                            className={`${prefix}--bmrg-privacy-statement__content`}
                            dangerouslySetInnerHTML={{
                              __html: dompurify.sanitize(section.content),
                            }}
                          />
                        </AccordionItem>
                      );
                    })}
                  </Accordion>
                  <p className={`${prefix}--bmrg-privacy-statement__message`}>
                    {`For any questions or concerns about business and personal information captured on
                  ${organization}, please contact${" "}`}
                    <a href={`mailto:${platformEmail}?subject=${organization} Privacy Statement`}>{platformEmail}</a>.
                  </p>
                </>
              )
            )}

            {mutateUserConsentError && (
              <p className={`${prefix}--bmrg-privacy-statement__error`}>
                Failed to receive deletion request. Please try again.
              </p>
            )}  
          </div>
        </ModalBody>
        <ModalFooter>
          <div className={`${prefix}--bmrg-privacy-statement-delete`}>
            <Button data-modal-primary-focus kind="danger" onClick={() => setIsConfirmModalOpen(true)}>
              Request account deletion
            </Button>
            <ComposedModal open={isConfirmModalOpen}>
              <ModalHeader
                closeModal={() => setIsConfirmModalOpen(false)}
                label="Delete Account"
                title="Request account deletion"
              />
              <ModalBody>
                <p className={`${prefix}--bmrg-privacy-statement-delete__desc`}>
                  By selecting to delete your account, your account will be deleted along with all of your user data
                  from our system and we will notify your team(s) that you are no longer a memeber of the platform. Are
                  you sure you want to delete your account?
                </p>
              </ModalBody>
              <ModalFooter style={{ marginTop: "1.125rem" }}>
                <Button data-modal-primary-focus kind="secondary" onClick={() => setIsConfirmModalOpen(false)}>
                  No, go back to Privacy Statement
                </Button>
                <Button
                  kind="danger"
                  type="submit"
                  onClick={() => {
                    handleSubmit({ closeAlertModal: () => setIsConfirmModalOpen(false), closeModal });
                  }}
                >
                  Yes, delete my account
                </Button>
              </ModalFooter>
            </ComposedModal>
          </div>
        </ModalFooter>
      </ComposedModal>
  );
}

function PrivacyStatementMenuItem(props: Omit<Props, "isOpen" | "closeModal">) {
  const menuItemRef = React.useRef<HTMLLinkElement>(null);
  const [isOpen, setIsOpen] = React.useState(false);

  const handleClose = () => {
    setIsOpen(false);
    setTimeout(() => {
      menuItemRef.current?.focus();
    }, 0);
  };

  return (
    <>
      <HeaderMenuItem
        icon={<Locked />}
        onClick={() => setIsOpen(!isOpen)}
        ref={menuItemRef}
        text="Privacy Statement"
        type="button"
      />
      <PrivacyStatement isOpen={isOpen} closeModal={handleClose} {...props} />
    </>
  );
}

export default PrivacyStatement;

export { PrivacyStatementMenuItem };
