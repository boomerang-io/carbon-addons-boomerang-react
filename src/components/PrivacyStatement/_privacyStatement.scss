@include exports('addons-bmrg-privacyStatement') {
  .#{$prefix}--bmrg-privacy-statement-container {
    .#{$prefix}--modal-container {
      height: rem(598px);
      width: rem(700px);
    }

    .#{$prefix}--bmrg-privacy-statement-delete {
      width: 100%;

      .#{$prefix}--modal-container {
        height: rem(476px);
        width: rem(500px);
        padding-left: 0;
        padding-right: 0;
      }

      .#{$prefix}--modal-content {
        overflow-x: hidden;
      }

      .#{$prefix}--modal-header {
        padding-left: 2rem;
      }

      .#{$prefix}--modal-footer {
        margin-top: 1.125rem;
      }

      .#{$prefix}--btn--danger {
        width: 100%;
      }

      .#{$prefix}--bmrg-privacy-statement-delete__desc {
        width: rem(420px);
        color: #9fa5ad;
        font-family: 'IBM Plex Sans', 'Helvetica Neue', Arial, Helvetica, sans-serif;
        font-size: rem(14px);
        line-height: rem(18px);
        padding-left: 2rem;
        margin-top: 1rem;
      }
    }

    .#{$prefix}--modal-content {
      .#{$prefix}--bmrg-privacy-statement {
        height: rem(368px);
        margin-top: rem(16px);
        overflow-y: auto;

        .#{$prefix}--accordion {
          overflow-y: hidden;
        }

        &__content {
          h3 {
            color: $bmrg-brand-white;
            padding: 1rem 0 0.3125rem;
          }
          a {
            color: $bmrg-secondary-color;
            text-decoration: none;
          }
        }

        &__message {
          height: rem(36px);
          width: rem(620px);
          color: #9fa5ad;
          font-size: rem(14px);
          line-height: rem(18px);
          margin-top: rem(24px);

          a {
            color: $bmrg-secondary-color;
            text-decoration: none;
          }
        }
        &__error {
          color: #fb4b53;
          font-size: 0.875rem;
          font-weight: 400;
          line-height: 1.125rem;
          margin-top: rem(4px);
          position: absolute;
        }
      }

      .#{$prefix}--accordion__arrow {
        transform: rotate(90deg);
        path {
          fill: $bmrg-brand-white;
        }
      }

      .#{$prefix}--accordion__item {
        border-top-color: rgba(251, 252, 252, 0.1);
        &:last-child {
          border-bottom-color: rgba(251, 252, 252, 0.1);
        }
        &--active {
          .#{$prefix}--accordion__heading > .#{$prefix}--accordion__arrow {
            transform: rotate(-90deg);
          }
        }
      }

      .#{$prefix}--accordion__heading:focus {
        .#{$prefix}--accordion__arrow {
          outline: rem(1px) solid $bmrg-secondary-color;
        }
      }

      .#{$prefix}--accordion__title {
        height: rem(18px);
        color: $bmrg-brand-white;
        font-size: rem(14px);
        font-weight: 600;
        line-height: rem(18px);
      }

      .#{$prefix}--accordion__content {
        color: #9fa5ad;
        font-family: 'IBM Plex Sans', 'Helvetica Neue', Arial, Helvetica, sans-serif;
        font-size: rem(14px);
        line-height: rem(18px);
      }
    }
  }
}
