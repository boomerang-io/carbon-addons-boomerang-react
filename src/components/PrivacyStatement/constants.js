export const PRIVACY_DATA = {
  id: '5af1c4545755960f83a7a671',
  version: 1,
  title: 'Privacy Statement',
  effectiveDate: '2019-08-14T16:17:45.354+0000',
  formContent: {
    sections: [
      {
        title: 'Introduction',
        content:
          "Your privacy is important to IBM; maintaining your trust is paramount to us. This Statement discloses the information practices for IBM’s Boomerang platform, what type of information about our websites’ users is gathered and tracked, and how the information is used, shared or otherwise processed offline. By accessing, viewing, and sharing content from IBM Boomerang, you are consenting to the collection and tracking of your personal data including your name, geography and IBM email address and the activities you specifically executed on IBM Boomerang. The data is collected as part of the user profile on the platform to help enable single sign-on during log-in. All other data is collected for usage and performance analytics. <h3>Website visits and IBM Cloud services</h3> We may also collect information relating to your use of our websites and IBM Cloud services through the use of various technologies. For example, when you visit our websites or access our IBM Cloud services, we may log certain information that your browser sends us, such as your IP address (including information deriving from your IP address such as your geographic location), browser type, version and language, access time, duration of access, and referring website addresses; we may also collect information about the pages you view within our sites, the time you spent on each site and other actions you take while visiting our website. In addition, some of our software products and IBM Cloud services include technologies that allow IBM to collect certain information about the use of our products and services. We may also use such technologies to determine whether you've opened an e-mail or clicked on a link contained in an e-mail. For details regarding the technologies we employ, see Cookies below. ",
      },
      {
        title: 'Retention',
        content:
          'IBM will retain your personal information for as long as is required to fulfill the purposes for which the information is processed or for other valid reasons to retain your personal information (for example to comply with our legal obligations, resolve disputes, or enforce our agreements). Your registration information will be retained for as long as your account or IBMid is active or as needed to provide you services. If you wish to request that we no longer use your registration information to provide you services, contact us at boomrng@us.ibm.com. ',
      },
      {
        title: 'Cookies',
        content:
          "As mentioned above, we collect information from your visits to our websites and your use of our IBM Cloud services and our software products to help us gather statistics about usage and effectiveness, personalize your experience, tailor our interactions with you, and improve our products and services. We do so through the use of various technologies, including scripts, tags, Local Shared Objects (Flash cookies), Local Storage (HTML5) beacons, and 'cookies'. <h3>What are cookies and why are cookies used</h3> A cookie is a piece of data that a website can send to your browser, which may then be stored on your computer as a tag that identifies your computer. While cookies are often only used to measure website usage (such as number of visitors and duration of visit) and effectiveness (such as topics visitors are most interested in) and to allow for ease of navigation or use and, as such, are not associated with any personal information, they are also used at times to personalize a known visitor's experience of a website by being associated with profile information or user preferences. Over time this information provides valuable insight to help improve the user experience. Cookies are typically categorized as 'session' cookies or 'persistent' cookies. Session cookies help you navigate through the website efficiently, keeping track of your progression from page to page so that you are not asked for information you have already provided during the current visit, or information needed to be able to complete a transaction. Session cookies are stored in temporary memory and erased when the web browser is closed. Persistent cookies on the other hand, store user preferences for current and successive visits. They are written on your device's hard disk, and are still valid when you restart your browser. We use persistent cookies, for example, to record your choice of language and country location. <h3>How to express privacy preferences regarding the use of cookies</h3> When visiting our websites or online services, you may have the possibility to set your preferences regarding cookies and other similar technologies by using the options and tools made available to you by either your web browser or IBM. If a cookie manager has been implemented by IBM, it will be displayed at your first visit to our website. By clicking ‘cookie preferences’ at the footer of an IBM website you can choose and change at any time which types of cookies you would like our websites and IBM Cloud services to use. You can indicate your preference for required, required and functional or required, functional and personalization cookies. Here you can also find more information on the types of cookies used. Using the cookie preferences does not result in deletion of cookies that are already on your device. You can delete cookies already on your device via the settings of your web browser. While IBM websites, online and Cloud services at this time do not recognize automated browser signals regarding tracking mechanisms, such as 'do not track' instructions, you can generally express your privacy preferences regarding the use of most cookies and similar technologies through your web browser. Look under the heading 'Tools' (or similar heading) in your particular browser for information about controlling cookies. In most instances you can set your browser to notify you before you receive a cookie, giving you the option to decide whether to accept it or not. You can also generally set your browser to turn off cookies. Cookies in our software products can be turned off in the product itself. Since cookies allow you to take advantage of some of our websites' features or features of our software products and IBM Cloud services, we recommend that you leave them turned on. If you block, turn off or otherwise reject our cookies, some web pages may not display properly or you will not be able to",
      },
      {
        title: 'Links to Non-IBM Websites and Third-Party Applications',
        content:
          'We may also provide you with general links to non-IBM websites. Your use of these links and applications is subject to the third parties’ privacy policies, and you should become familiar with the third-party sites\' privacy policies before using the links or applications. IBM is not responsible for the privacy practices or the content of those other websites. Individuals with access to this information must adhere to all applicable IBM policies, practices and guidelines regarding the protection of personal information and IBM Confidential information. They must not further disseminate the data to anyone without a legitimate need-to-know. They must also not use this information for any other purpose than those listed above. For any questions or concerns about business personal information captured on IBM Boomerang, please contact boomrng@us.ibm.com. Additionally IBM’s Global Online Privacy Statement can be found <a href="ibm.com">here</a>. By clicking on “I understand” you are confirming that you have read the IBM Boomerang privacy statement, that you understand it, and that you agree to it. If you do not agree to the personal data collection and tracking from the IBM Boomerang, you should exit out of the hub. ',
      },
    ],
  },
};
