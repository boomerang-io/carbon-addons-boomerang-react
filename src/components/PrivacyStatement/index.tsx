export { default, PrivacyStatementMenuItem } from "./PrivacyStatement";
