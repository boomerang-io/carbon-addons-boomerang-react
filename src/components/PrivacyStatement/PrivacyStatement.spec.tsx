import React from "react";
import { expect, test, vi } from "vitest";
import { QueryClient, QueryClientProvider } from "react-query";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { serviceUrl } from "../../config/servicesConfig";
import PrivacyStatement from "./PrivacyStatement";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { PRIVACY_DATA } from "./constants";

const queryClient = new QueryClient({
  defaultOptions: { queries: { refetchOnWindowFocus: false } },
});

const baseServiceUrl = "http://boomerang.com";
const { reload } = window.location;

beforeAll(() => {
  Object.defineProperty(window, "location", {
    writable: true,
    value: { reload: vi.fn() },
  });
});

afterAll(() => {
  window.location.reload = reload;
});

test("Privacy Statement error", async () => {
  /**
   * Simulate deleting account
   * -pulls in mocked out data
   * -ends in failure based on
   *
   * NOTE: there is a conole warning here stating that when updating the state
   * of a component, you should use the act() function. Bug in React Hooks:
   * https://github.com/kentcdodds/@testing-library/react/issues/285
   *
   */
  const mock = new MockAdapter(axios);
  mock.onGet(serviceUrl.getStatement({ baseServiceUrl })).reply(200, PRIVACY_DATA);
  mock.onPut(serviceUrl.resourceUserConsent({ baseServiceUrl })).networkError();
  const { getByText, getByRole, findByRole } = render(
    <QueryClientProvider client={queryClient}>
      <PrivacyStatement baseServiceUrl={baseServiceUrl} />
    </QueryClientProvider>
  );
  const btn = getByRole("button", { name: /^Privacy Statement$/i });
  fireEvent.click(btn);
  const deleteButton = await findByRole("button", { name: /Request account deletion/i });
  fireEvent.click(deleteButton);
  const confirmButton = getByText(/Delete my account/i);
  fireEvent.click(confirmButton);
  await waitFor(() =>
    (expect(getByText(/Failed to receive deletion request. Please try again./i)) as any).toBeInTheDocument()
  );
});

test("Privacy Statement success", async () => {
  /**
   *
   */
  const mock = new MockAdapter(axios);
  mock.onGet(`${baseServiceUrl}/users/consents`).reply(200, PRIVACY_DATA);
  mock.onPut(`${baseServiceUrl}/users/consent`).reply(200);

  const { getByRole, findByRole } = render(
    <QueryClientProvider client={queryClient}>
      <PrivacyStatement baseServiceUrl={baseServiceUrl} />
    </QueryClientProvider>
  );

  const btn = getByRole("button", { name: /^Privacy Statement$/i });
  fireEvent.click(btn);

  const deleteButton = await findByRole("button", { name: /Request account deletion/i });
  fireEvent.click(deleteButton);

  const confirmButton = getByRole("button", { name: /Delete my account/i });
  fireEvent.click(confirmButton);
});
