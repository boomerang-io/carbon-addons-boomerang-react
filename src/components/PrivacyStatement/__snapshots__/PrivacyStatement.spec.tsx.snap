// Vitest Snapshot v1

exports[`Privacy Statement - snapshot 1`] = `
<body
  class="cds--body--with-modal-open"
>
  <div>
    <div
      aria-hidden="false"
      class="cds--modal is-visible cds--bmrg-privacy-statement-container cds--bmrg-header-modal"
      role="presentation"
      tabindex="-1"
    >
      <div
        aria-label="Privacy Statement"
        aria-modal="true"
        class="cds--modal-container"
        role="dialog"
      >
        <button
          class="cds--visually-hidden"
          type="button"
        >
          Focus sentinel
        </button>
        <div
          class="cds--modal-container-body"
        >
          <div
            class="cds--modal-header"
          >
            <h2
              class="cds--modal-header__label cds--type-delta"
            >
              Effective as of 
            </h2>
            <h3
              class="cds--modal-header__heading cds--type-beta"
            >
              Privacy Statement
            </h3>
            <button
              aria-label="Close"
              class="cds--modal-close"
              title="Close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="cds--modal-close__icon"
                fill="currentColor"
                focusable="false"
                height="20"
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                />
              </svg>
            </button>
          </div>
          <div
            class="cds--modal-content"
          >
            <div
              class="cds--bmrg-privacy-statement"
            />
          </div>
          <div
            class="cds--modal-footer cds--btn-set"
          >
            <button
              class="cds--btn cds--btn--secondary"
              data-modal-primary-focus="true"
              tabindex="0"
              type="button"
            >
              Close
            </button>
            <button
              aria-describedby="danger-description-2"
              class="cds--btn cds--btn--danger"
              tabindex="0"
              type="button"
            >
              <span
                class="cds--visually-hidden"
                id="danger-description-2"
              >
                danger
              </span>
              Request account deletion
            </button>
            <div
              class="cds--bmrg-privacy-statement-delete"
            >
              <div
                aria-hidden="true"
                class="cds--modal"
                role="presentation"
                tabindex="-1"
              >
                <div
                  aria-label="Delete Account"
                  aria-modal="true"
                  class="cds--modal-container"
                  role="dialog"
                >
                  <button
                    class="cds--visually-hidden"
                    type="button"
                  >
                    Focus sentinel
                  </button>
                  <div
                    class="cds--modal-container-body"
                  >
                    <div
                      class="cds--modal-header"
                    >
                      <h2
                        class="cds--modal-header__label cds--type-delta"
                      >
                        Delete Account
                      </h2>
                      <h3
                        class="cds--modal-header__heading cds--type-beta"
                      >
                        Request account deletion
                      </h3>
                      <button
                        aria-label="Close"
                        class="cds--modal-close"
                        title="Close"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="cds--modal-close__icon"
                          fill="currentColor"
                          focusable="false"
                          height="20"
                          preserveAspectRatio="xMidYMid meet"
                          viewBox="0 0 32 32"
                          width="20"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
                          />
                        </svg>
                      </button>
                    </div>
                    <div
                      class="cds--modal-content"
                    >
                      <p
                        class="cds--bmrg-privacy-statement-delete__desc"
                      >
                        By selecting to delete your account, your account will be deleted along with all of your user data from our system and we will notify your team(s) that you are no longer a memeber of the platform. Are you sure you want to delete your account?
                      </p>
                    </div>
                    <div
                      class="cds--modal-footer cds--btn-set"
                      style="margin-top: 1.125rem;"
                    >
                      <button
                        class="cds--btn cds--btn--secondary"
                        data-modal-primary-focus="true"
                        tabindex="0"
                        type="button"
                      >
                        No, go back to Privacy Statement
                      </button>
                      <button
                        aria-describedby="danger-description-4"
                        class="cds--btn cds--btn--danger"
                        tabindex="0"
                        type="submit"
                      >
                        <span
                          class="cds--visually-hidden"
                          id="danger-description-4"
                        >
                          danger
                        </span>
                        Yes, delete my account
                      </button>
                    </div>
                  </div>
                  <button
                    class="cds--visually-hidden"
                    type="button"
                  >
                    Focus sentinel
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <button
          class="cds--visually-hidden"
          type="button"
        >
          Focus sentinel
        </button>
      </div>
    </div>
  </div>
</body>
`;
